wpilibc/src/main/native/cpp/apriltag/AprilTagFieldLayout.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include "frc/apriltag/AprilTagFieldLayout.h"
+
+#include <algorithm>
+#include <system_error>
+
+#include <units/angle.h>
+#include <units/length.h>
+#include <wpi/json.h>
+#include <wpi/raw_istream.h>
+#include <wpi/raw_ostream.h>
+
+using namespace frc;
+
+AprilTagFieldLayout::AprilTagFieldLayout(std::string_view path) {
+  std::error_code error_code;
+
+  wpi::raw_fd_istream input{path, error_code};
+  if (error_code) {
+    throw std::runtime_error(fmt::format("Cannot open file: {}", path));
+  }
+
+  wpi::json json;
+  input >> json;
+
+  m_apriltags = json.get<std::vector<AprilTag>>();
+}
+
+AprilTagFieldLayout::AprilTagFieldLayout(std::vector<AprilTag> apriltags)
+    : m_apriltags(std::move(apriltags)) {}
+
+void AprilTagFieldLayout::SetAlliance(DriverStation::Alliance alliance) {
+  m_mirror = alliance == DriverStation::Alliance::kRed;
+}
+
+frc::Pose3d AprilTagFieldLayout::GetTagPose(int ID) const {
+  Pose3d returnPose;
+  auto it = std::find_if(m_apriltags.begin(), m_apriltags.end(),
+                         [=](const auto& tag) { return tag.ID == ID; });
+  if (it != m_apriltags.end()) {
+    returnPose = it->pose;
+  }
+  if (m_mirror) {
+    returnPose = returnPose.RelativeTo(
+        Pose3d{54_ft, 27_ft, 0_ft, Rotation3d{0_deg, 0_deg, 180_deg}});
+  }
+  return returnPose;
+}
+
+void AprilTagFieldLayout::Serialize(std::string_view path) {
+  std::error_code error_code;
+
+  wpi::raw_fd_ostream output{path, error_code};
+  if (error_code) {
+    throw std::runtime_error(fmt::format("Cannot open file: {}", path));
+  }
+
+  wpi::json json = *this;
+  output << json;
```

Comment:
Will this append if the file already exists?

Reply:
It truncates the file if it already exists (i.e., it clears the previous contents).
```cpp
raw_fd_ostream::raw_fd_ostream(std::string_view Filename, std::error_code &EC)
    : raw_fd_ostream(Filename, EC, fs::CD_CreateAlways, fs::FA_Write,
                     fs::OF_None) {}
```
```cpp
enum CreationDisposition : unsigned {
  /// CD_CreateAlways - When opening a file:
  ///   * If it already exists, truncate it.
  ///   * If it does not already exist, create a new file.
  CD_CreateAlways = 0,
```